### PR TITLE
[Feature] Storage volume RBAC

### PR DESCRIPTION
**Pull request by TszKitLo40**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 123 (56.10%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/system/starrocks/GrantsTo.java|8|20|40.00%|[403, 404, 405, 406, 409, 410, 411, 413, 415, 458, 459, 460]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|21|43|48.84%|[473, 474, 475, 477, 1231, 1232, 1233, 1234, 1235, 1236, 1237, 1801, 1811, 1821, 1829, 1830, 1831, 1834, 1840, 1841, 1842, 1845]|
|:large_blue_circle:|com/starrocks/privilege/StorageVolumePEntryObject.java|24|44|54.55%|[35, 44, 45, 48, 61, 78, 79, 80, 86, 87, 89, 91, 92, 93, 95, 98, 99, 101, 107, 110]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/ObjectType.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeActions.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|2|2|100.00%|[]|

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 66 / 122 (54.10%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/system/starrocks/GrantsTo.java|8|20|40.00%|[403, 404, 405, 406, 409, 410, 411, 413, 415, 458, 459, 460]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|17|41|41.46%|[473, 474, 475, 477, 697, 1231, 1232, 1233, 1234, 1235, 1236, 1237, 1244, 1801, 1811, 1821, 1829, 1830, 1831, 1834, 1840, 1841, 1842, 1845]|
|:large_blue_circle:|com/starrocks/privilege/StorageVolumePEntryObject.java|25|45|55.56%|[35, 44, 45, 48, 61, 78, 79, 80, 86, 87, 89, 91, 92, 93, 95, 98, 99, 101, 107, 110]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/ObjectType.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeActions.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java [hunk 1]
```diff
@@ -97,6 +98,11 @@ public class DefaultAuthorizationProvider implements AuthorizationProvider {
                     .put(ObjectType.GLOBAL_FUNCTION, ImmutableList.of(
                             PrivilegeType.USAGE,
                             PrivilegeType.DROP))
+
+                    .put(ObjectType.STORAGE_VOLUME, ImmutableList.of(
+                            PrivilegeType.DROP,
+                            PrivilegeType.ALTER,
+                            PrivilegeType.USAGE))
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java:105]: which operation should check usage on storage volume?

File: fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java [hunk 2]
```diff
@@ -155,6 +161,9 @@ public PEntryObject generateObject(ObjectType objectType, List<String> objectTok
             case RESOURCE_GROUP:
                 return ResourceGroupPEntryObject.generate(mgr, objectTokens);
 
+            case STORAGE_VOLUME:
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java:164]: built-in role `db_admin` should have any priv on any storage volumn?

File: fe/fe-core/src/main/java/com/starrocks/privilege/StorageVolumePEntryObject.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            throw new PrivilegeException("invalid object tokens, should have one: " + tokens);
+        }
+        String name = tokens.get(0);
+        if (name.equals("*")) {
+            return new StorageVolumePEntryObject(null);
+        } else {
+            StorageVolume sv = null;
+            try {
+                sv = mgr.getStorageVolumeMgr().getStorageVolumeByName(name);
+            } catch (AnalysisException e) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/StorageVolumePEntryObject.java:44]: why getStorageVolumeByName throws `AnalysisException`? is this reasonable?

File: fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java [hunk 4]
```diff
@@ -2553,8 +2553,9 @@ private void handleShowStorageVolumes() throws DdlException {
         }
         PatternMatcher finalMatcher = matcher;
         storageVolumeNames = storageVolumeNames.stream()
-                .filter(tblName -> finalMatcher == null || finalMatcher.match(tblName))
+                .filter(storageVolumeName -> finalMatcher == null || finalMatcher.match(storageVolumeName))
                 // TODO: check privilege
+                .filter(storageVolumeName -> PrivilegeActions.checkAnyActionOnStorageVolume(connectContext, storageVolumeName))
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java:2558]: remove todo

File: fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java (hunk 1 above)
Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java:105]: Create db/table using a storage volume

File: fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java (hunk 2 above)
Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/privilege/DefaultAuthorizationProvider.java:164]: > built-in role `db_admin` should have any priv on any storage volumn?

It should have all of the priv on storage volumes

File: fe/fe-core/src/main/java/com/starrocks/privilege/StorageVolumePEntryObject.java (hunk 3 above)
Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/privilege/StorageVolumePEntryObject.java:44]: I will refactor the exception thrown in another PR.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 66 / 122 (54.10%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/system/starrocks/GrantsTo.java|8|20|40.00%|[403, 404, 405, 406, 409, 410, 411, 413, 415, 458, 459, 460]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|17|41|41.46%|[473, 474, 475, 477, 697, 1231, 1232, 1233, 1234, 1235, 1236, 1237, 1244, 1801, 1811, 1821, 1829, 1830, 1831, 1834, 1840, 1841, 1842, 1845]|
|:large_blue_circle:|com/starrocks/privilege/StorageVolumePEntryObject.java|25|45|55.56%|[35, 45, 46, 49, 62, 79, 80, 81, 87, 88, 90, 92, 93, 94, 96, 99, 100, 102, 108, 111]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/ObjectType.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeActions.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|2|2|100.00%|[]|

**Review by nshangyiming — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**